musicbrainz/graphql.go

```diff
@@ -0,0 +1,201 @@
+// This file is part of the go-meta library.
+//
+// Copyright (C) 2017 JAAK MUSIC LTD
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+// If you have any questions please contact yo@jaak.io
+
+package musicbrainz
+
+import (
+	"database/sql"
+	"errors"
+
+	"github.com/ipfs/go-cid"
+	"github.com/meta-network/go-meta"
+)
+
+// GraphQLSchema is the GraphQL schema for the MusicBrainz META index.
+const GraphQLSchema = `
+schema {
+  query: Query
+}
+
+type Query {
+  artist(
+    name: String,
+    ipi:  String,
+    isni: String
+  ): [Artist]!
+}
+
+type Artist {
+  cid:                    String!
+  name:                   String!
+  sort_name:              String!
+  type:                   String
+  gender:                 String
+  area:                   String
+  begin_date:             String
+  end_date:               String
+  ipi:                    [String!]
+  isni:                   [String!]
+  alias:                  [String!]
+  mbid:                   String!
+  disambiguation_comment: String
+  annotation:             [String!]
+}
+`
+
+// Resolver defines GraphQL resolver functions for the schema contained in
+// the GraphQLSchema constant, retrieving data from a META store and SQLite3
+// index.
+type Resolver struct {
+	db    *sql.DB
+	store *meta.Store
+}
+
+// NewResolver returns a Resolver which retrieves data from the given META
+// store and SQLite3 index.
+func NewResolver(db *sql.DB, store *meta.Store) *Resolver {
+	return &Resolver{db, store}
+}
+
+// artistArgs are the arguments for a GraphQL artist query.
+type artistArgs struct {
+	Name *string
+	IPI  *string
+	ISNI *string
+}
+
+// Artist is a GraphQL resolver function which retrieves object IDs from the
+// SQLite3 index using either an artist name, IPI or ISNI, and loads the
+// associated META objects from the META store.
+func (g *Resolver) Artist(args artistArgs) ([]*artistResolver, error) {
+	var rows *sql.Rows
+	var err error
+	switch {
+	case args.Name != nil:
+		rows, err = g.db.Query("SELECT object_id FROM artist WHERE name = ?", *args.Name)
```

Comment:
If I specify `Name` AND (`IPI` AND/OR `ISNI`), name takes precedence. Is this desired behavior?

Reply:
This API accepts *either* a Name, IPI or ISNI, so this is the desired behaviour yes.